File: typescript/packages/core/src/http.ts
import { anyAbortSignal } from './abort.js';
import {
    createSignerError,
    sanitizeRemoteErrorResponse,
    SignerError,
    SignerErrorCode,
    throwSignerError,
} from './errors.js';

/** Default timeout applied to remote signer API requests. */
export const DEFAULT_FETCH_TIMEOUT_MS = 60_000;

/**
 * Maximum number of response-body bytes accepted from a remote signer API
 * (1 MiB, matching the Go backend). Larger bodies fail with `PARSING_ERROR`
 * instead of being buffered unbounded.
 */
export const MAX_RESPONSE_BYTES = 1024 * 1024;

const providerResponseErrors = new WeakSet<SignerError>();

function throwProviderResponseError(code: SignerErrorCode, context: Record<string, unknown>): never {
    const error = createSignerError(code, context);
    providerResponseErrors.add(error);
    throw error;
}

export interface FetchSignerJsonOptions {
    /**
     * Caller-supplied cancellation, typically the `abortSignal` of a Kit signer
     * config. It composes with the timeout: whichever fires first aborts the
     * request. When it has already fired, no request is made at all and the
     * abort reason is thrown unwrapped rather than as a signer error.
     */
    abortSignal?: AbortSignal;
    /**
     * Standard fetch options (method, headers, body, ...).
     *
     * A `signal` here is the raw escape hatch for callers that own the whole
     * request lifetime: it replaces the timeout instead of composing with it,
     * and still composes with {@link FetchSignerJsonOptions.abortSignal}.
     * Prefer {@link FetchSignerJsonOptions.abortSignal} for cancellation.
     */
    init?: RequestInit;
    /** Human-readable provider name used in error messages (e.g. "Privy"). */
    providerName: string;
    /**
     * Per-request timeout in ms. Ignored when `init.signal` is provided.
     * Default: {@link DEFAULT_FETCH_TIMEOUT_MS}.
     */
    timeoutMs?: number;
    /** Fully-qualified request URL. */
    url: string;
}

/** The provider's HTTP status when its response was the failure. */
export function providerStatus(error: unknown): number | undefined {
    const status =
        error instanceof SignerError && providerResponseErrors.has(error) ? error.context?.status : undefined;
    return typeof status === 'number' ? status : undefined;
}

/**
 * A 4xx other than 408 is the only create outcome that rules out a transaction;
 * anything else (no response, timeout, 5xx, unusable success body) may already be
 * executing. A 408 is a timeout reached while the request was being processed, so
 * it does not rule the transaction out either.
 */
export function providerMayHaveAccepted(error: unknown): boolean {
    const status = providerStatus(error);
    return status === undefined || status < 400 || status >= 500 || status === 408;
}

function throwResponseTooLarge(providerName: string, status: number): never {
    throwProviderResponseError(SignerErrorCode.PARSING_ERROR, {
        maxResponseBytes: MAX_RESPONSE_BYTES,
        message: `${providerName} response exceeded maximum size`,
        status,
    });
}

/**
 * Read a response body as text while enforcing {@link MAX_RESPONSE_BYTES}.
 *
 * A `Content-Length` header over the cap fails fast, but the cap is always
 * enforced while streaming too (the header can lie or be absent). Returns
 * `undefined` when the response exposes no readable body stream, so callers
 * can fall back to the non-streaming read.
 */
async function readCappedResponseText(response: Response, providerName: string): Promise<string | undefined> {
    const contentLength = Number(response.headers?.get('content-length'));
    if (Number.isFinite(contentLength) && contentLength > MAX_RESPONSE_BYTES) {
        throwResponseTooLarge(providerName, response.status);
    }

    const body = response.body;
    if (!body) {
        return undefined;
    }

    const reader = body.getReader();
    const chunks: Uint8Array[] = [];
    let totalBytes = 0;
    try {
        for (;;) {
            const { done, value } = await reader.read();
            if (done) break;
            if (!value) continue;
            totalBytes += value.byteLength;
            if (totalBytes > MAX_RESPONSE_BYTES) {
                throwResponseTooLarge(providerName, response.status);
            }
            chunks.push(value);
        }
    } finally {
        await reader.cancel().catch(() => {});
    }

    const bytes = new Uint8Array(totalBytes);
    let offset = 0;
    for (const chunk of chunks) {
        bytes.set(chunk, offset);
        offset += chunk.byteLength;
    }
    return new TextDecoder().decode(bytes);
}

/**
 * The top-level `id` of a failed response body, when there is one. A provider
 * that has already accepted a transaction may still answer with a non-2xx
 * status, and that id is the caller's only handle for reconciling it.
 */
function transactionIdInBody(body: string): string | undefined {
    let parsed: unknown;
    try {
        parsed = JSON.parse(body);
    } catch {
        return undefined;
    }
    if (typeof parsed !== 'object' || parsed === null) return undefined;
    const transactionId = (parsed as { id?: unknown }).id;
    return typeof transactionId === 'string' && transactionId.trim() ? transactionId : undefined;
}

/**
 * Perform a remote signer API request and parse the JSON response, mapping
 * failures to the standard signer error pipeline:
 * - network failure or timeout → `HTTP_ERROR`
 * - non-2xx status → `REMOTE_API_ERROR` with the sanitized response body
 * - invalid JSON body → `PARSING_ERROR`
 * - body over {@link MAX_RESPONSE_BYTES} → `PARSING_ERROR`
 *
 * Redirects are always rejected and every request carries a timeout unless
 * the caller supplies its own `init.signal`. A caller `abortSignal` propagates
 * its abort reason unwrapped, so cancellation is distinguishable from failure.
 *
 * A non-2xx response is the one exception: the provider has already returned
 * its verdict, and that verdict is what callers classify broadcast safety on
 * (a 4xx rules the transaction out, an abort reason does not). Such a response
 * surfaces as `REMOTE_API_ERROR` carrying the status even when the caller
 * aborts while the error body is being read.
 */
export async function fetchSignerJson<TResponse>(options: FetchSignerJsonOptions): Promise<TResponse> {
    const { abortSignal, init = {}, providerName, timeoutMs = DEFAULT_FETCH_TIMEOUT_MS, url } = options;

    abortSignal?.throwIfAborted();

    const signals: AbortSignal[] = [];
    if (abortSignal) signals.push(abortSignal);
    signals.push(init.signal ?? AbortSignal.timeout(timeoutMs));

    let response: Response;
    try {
        response = await fetch(url, {
            ...init,
            redirect: 'error',
            signal: anyAbortSignal(signals),
        });
    } catch (error) {
        abortSignal?.throwIfAborted();
        throwSignerError(SignerErrorCode.HTTP_ERROR, {
            cause: error,
            message: `${providerName} network request failed`,
            url,
        });
    }

    if (!response.ok) {
        let errorText: string;
        try {
            errorText = (await readCappedResponseText(response, providerName)) ?? (await response.text());
        } catch (error) {
            // An over-cap body is a hard failure; any other read failure keeps
            // the provider's verdict with a placeholder body.
            if (error instanceof SignerError) throw error;
            errorText = 'Failed to read error response';
        }
        const providerTransactionId = transactionIdInBody(errorText);
        throwProviderResponseError(SignerErrorCode.REMOTE_API_ERROR, {
            message: `${providerName} API error: ${response.status}`,
            ...(providerTransactionId === undefined ? {} : { providerTransactionId }),
            response: sanitizeRemoteErrorResponse(errorText),
            status: response.status,
        });
    }

    try {
        const text = await readCappedResponseText(response, providerName);
        return (text === undefined ? await response.json() : JSON.parse(text)) as TResponse;
    } catch (error) {
        // The over-cap error is already a fully-classified PARSING_ERROR.
        if (error instanceof SignerError) throw error;
        abortSignal?.throwIfAborted();
        throwSignerError(SignerErrorCode.PARSING_ERROR, {
            cause: error,
            message: `Failed to parse ${providerName} response`,
        });
    }
}

File: python/src/solana_keychain/core/http.py
"""HTTPS-enforcing HTTP pipeline for remote signer backends."""

import asyncio
import json
import os
import re
from collections.abc import Awaitable, Callable
from typing import Any
from urllib.parse import urlsplit

import httpx

from solana_keychain.core.errors import SignerError, SignerErrorCode

DEFAULT_REQUEST_TIMEOUT_SECONDS = 60.0
AVAILABILITY_TIMEOUT_SECONDS = 5.0
DEFAULT_REMOTE_ERROR_RESPONSE_MAX_LENGTH = 256
MAX_RESPONSE_BYTES = 1024 * 1024

_LOOPBACK_HOSTNAMES = frozenset({"localhost", "127.0.0.1", "::1"})
_CONTROL_CHARS = re.compile(r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]")


def normalize_base_url(base_url: str) -> str:
    """Trim surrounding whitespace and strip trailing slashes so paths can be
    appended with a single ``/``."""
    return base_url.strip().rstrip("/")


def assert_https_url(url: str, field: str, *, allow_http_loopback_in_tests: bool = False) -> None:
    """Validate that a configured endpoint is a well-formed HTTPS URL — the security
    control behind the library's "HTTPS enforced" guarantee; every remote backend must
    run its base URL through this check.

    ``allow_http_loopback_in_tests`` permits plain-HTTP loopback URLs
    (localhost / 127.0.0.1 / ::1) while running under pytest, for backends whose
    integration tests target a local dev server.
    """
    try:
        parsed = urlsplit(url)
        hostname = parsed.hostname
        # Accessing ``port`` forces urllib to validate the authority's port syntax
        # and range rather than deferring the failure to httpx.
        _ = parsed.port
    except ValueError:
        raise SignerError(SignerErrorCode.CONFIG_ERROR, f"{field} is not a valid URL") from None
    if not hostname:
        raise SignerError(SignerErrorCode.CONFIG_ERROR, f"{field} is not a valid URL")
    if parsed.scheme == "https":
        return
    if (
        allow_http_loopback_in_tests
        and parsed.scheme == "http"
        and hostname in _LOOPBACK_HOSTNAMES
        and "PYTEST_CURRENT_TEST" in os.environ
    ):
        return
    raise SignerError(SignerErrorCode.CONFIG_ERROR, f"{field} must use HTTPS")


def sanitize_remote_error_response(
    response_text: str, max_length: int = DEFAULT_REMOTE_ERROR_RESPONSE_MAX_LENGTH
) -> str:
    """Sanitize remote API error text before attaching it to error detail: strips
    control characters, collapses whitespace, truncates long payloads."""
    normalized = " ".join(_CONTROL_CHARS.sub(" ", response_text).split())
    if not normalized:
        return "[empty remote response]"
    if len(normalized) <= max_length:
        return normalized
    return f"{normalized[:max_length]} [truncated]"


async def probe_availability(probe: Callable[[], Awaitable[bool]]) -> bool:
    """Run a backend health probe under the shared availability timeout.

    Any failure means unavailable: a signer error, a timeout, or an exception from
    caller-supplied credential machinery all report ``False`` rather than raising,
    since the contract returns a bool. Cancellation still propagates.
    """
    try:
        return await asyncio.wait_for(probe(), AVAILABILITY_TIMEOUT_SECONDS)
    except Exception:
        return False


def provider_may_have_accepted(status_code: int | None) -> bool:
    """A 4xx other than 408 is the only create outcome that rules out a transaction;
    anything else (no response, timeout, 5xx, unusable success body) may already be
    executing. A 408 is a timeout reached while the request was being processed, so it
    does not rule the transaction out either.
    """
    return status_code is None or status_code == 408 or not 400 <= status_code < 500


async def fetch_signer_json(
    *,
    url: str,
    provider_name: str,
    method: str = "GET",
    headers: dict[str, str] | None = None,
    json_body: Any | None = None,
    content: bytes | None = None,
    timeout_seconds: float = DEFAULT_REQUEST_TIMEOUT_SECONDS,
    client: httpx.AsyncClient | None = None,
) -> Any:
    """Perform a remote signer API request and parse the JSON response, mapping
    failures to the standard signer error pipeline:

    - network failure or timeout → ``HTTP_ERROR``
    - any redirect → ``HTTP_ERROR`` (auth headers must never replay against a
      redirect target)
    - non-2xx status → ``REMOTE_API_ERROR`` with the sanitized response body in
      the (redacted) detail
    - invalid JSON body → ``PARSING_ERROR``
    - body larger than ``MAX_RESPONSE_BYTES`` → ``PARSING_ERROR``

    ``json_body`` and ``content`` are mutually exclusive. Use ``content`` when the
    request bytes must match a signature computed over them exactly (request-stamping
    auth schemes) — re-serialization would change the bytes.

    When ``client`` is provided the caller owns its lifecycle and transport policy
    (custom TLS, proxies); otherwise a one-shot client is used.
    """
    if json_body is not None and content is not None:
        raise SignerError(
            SignerErrorCode.CONFIG_ERROR, "json_body and content are mutually exclusive"
        )
    if client is not None:
        return await _request_json(
            client,
            url=url,
            provider_name=provider_name,
            method=method,
            headers=headers,
            json_body=json_body,
            content=content,
            timeout_seconds=timeout_seconds,
        )
    async with httpx.AsyncClient() as own_client:
        return await _request_json(
            own_client,
            url=url,
            provider_name=provider_name,
            method=method,
            headers=headers,
            json_body=json_body,
            content=content,
            timeout_seconds=timeout_seconds,
        )


async def _request_json(
    client: httpx.AsyncClient,
    *,
    url: str,
    provider_name: str,
    method: str,
    headers: dict[str, str] | None,
    json_body: Any | None,
    content: bytes | None,
    timeout_seconds: float,
) -> Any:
    request = client.build_request(
        method,
        url,
        headers=headers,
        json=json_body,
        content=content,
        timeout=timeout_seconds,
    )
    try:
        response = await client.send(request, stream=True, follow_redirects=False)
    except httpx.HTTPError as error:
        raise SignerError(
            SignerErrorCode.HTTP_ERROR, f"{provider_name} network request failed: {error}"
        ) from None
    try:
        if 300 <= response.status_code < 400:
            raise SignerError(
                SignerErrorCode.HTTP_ERROR, f"{provider_name} response was a redirect"
            )
        body = await _read_bounded_body(response, provider_name)
    finally:
        await response.aclose()
    if not response.is_success:
        error_text = body.decode(response.encoding or "utf-8", errors="replace")
        raise SignerError(
            SignerErrorCode.REMOTE_API_ERROR,
            f"{provider_name} API error: {response.status_code}: "
            f"{sanitize_remote_error_response(error_text)}",
            provider_transaction_id=_transaction_id_in_body(body),
            status_code=response.status_code,
        )
    try:
        return json.loads(body)
    except ValueError:
        raise SignerError(
            SignerErrorCode.PARSING_ERROR, f"Failed to parse {provider_name} response"
        ) from None


def _transaction_id_in_body(body: bytes) -> str | None:
    """The top-level ``id`` of a failed response body, when there is one.

    A provider that has already accepted a transaction may still answer with a
    non-2xx status, and that id is the caller's only handle for reconciling it.
    """
    try:
        parsed = json.loads(body)
    except ValueError:
        return None
    if isinstance(parsed, dict):
        transaction_id = parsed.get("id")
        if isinstance(transaction_id, str) and transaction_id.strip():
            return transaction_id
    return None


async def _read_bounded_body(response: httpx.Response, provider_name: str) -> bytes:
    """Stream the response body, never buffering more than ``MAX_RESPONSE_BYTES``.

    Applies to success and error bodies alike so a misbehaving remote cannot
    exhaust memory before the status/JSON handling sees the payload.
    """
    body = bytearray()
    try:
        async for chunk in response.aiter_bytes():
            body.extend(chunk)
            if len(body) > MAX_RESPONSE_BYTES:
                raise SignerError(
                    SignerErrorCode.PARSING_ERROR,
                    f"{provider_name} response exceeded maximum size",
                )
    except httpx.HTTPError as error:
        raise SignerError(
            SignerErrorCode.HTTP_ERROR, f"{provider_name} network request failed: {error}"
        ) from None
    return bytes(body)
